model_training/distributed_training.py

```diff
@@ -0,0 +1,308 @@
+                                'total_msg': ..
+                            }
+                        """
+                        kafka_topic = data['topic']
+                        logging.info("Received control confirmation of data from Kafka for deployment ID %s. Ready to receive data from topic %s with batch %d", str(kafka_topic), deployment_id, batch)
+                        
+                        decoder = DecoderFactory.get_decoder(data['input_format'], data['input_config'])
+                        """Gets the decoder from the information received"""
+
+                        kafka_dataset = get_train_data(bootstrap_servers, kafka_topic, str(result_id), batch, decoder)
+                        """Gets the dataset from kafka"""
+
+                        end_consuming_data = time.time()
+                        logging.info("Total time consuming data: %s", str(end_consuming_data - start_consuming_data))
+
+                        logging.info("Model ready to be trained with configuration %s", str(kwargs_fit))
+                        
+                        split = int((1-data['validation_rate'])*(data['total_msg']/batch))
+                        validation_size= (data['total_msg']/batch)-split
+                        logging.info("Training batch size %d and validation batch size %d", split, validation_size)
+                        
+                        start_training = time.time()
+
+                        train_dataset = kafka_dataset.take(split).cache().repeat()
+                        """Splits dataset for training"""
+
+                        validation_dataset = kafka_dataset.skip(split).cache().repeat()
+                        """Splits dataset for validation"""
+                        
+                        logging.info("Splitting done, training is going to start.")
+
+                        """TENSORFLOW code goes here"""
+                        outputs = []
+                        img_input = tensorflow_models[0].input
+                        outputs.append(tensorflow_models[0](img_input))
+                        for index in range(1, N):
+                            next_input = outputs[index-1]
+                            outputs.append(tensorflow_models[index](next_input[0]))
+                            """Obteins all the outputs from each distributed submodel"""
+
+                        predictions = []
+                        for index in range(0, N-1):
+                            s = outputs[index]
+                            predictions.append(s[1])
+                        predictions.append(outputs[-1])
+                        """Obteins all the prediction outputs from each distributed submodel"""
+
+                        model = keras.Model(inputs=[img_input], outputs=predictions, name='model')
+                        """Creates a global model consisting of all distributed submodels"""
+
+                        weights = {}

```

Comment:
This variable is called `weights`, but stores the loss function name for each model.